test/Test/Database/LSMTree/Internal/Snapshot/Codec/Golden.hs

```diff
@@ -0,0 +1,255 @@
+        -- re-write out the serialized data. This ensures that there are no
+        -- false-positives, false-negatives, or irrelavent I/O exceptions.
+        removeIfExists snapshotFsPath
+        BSL.writeFile snapshotHsPath . toLazyByteString $ encode datum
+
+  in  Au.goldenVsFile name snapshotAuPath snapshotHsPath outputAction
+
+testCodecSnapshotLabel :: TestTree
+testCodecSnapshotLabel =
+  let assembler (tagA, valA) =
+        let (tagB, valB) = basicSnapshotTableType
+            (tagC, valC) = basicTableConfig
+            (tagD, valD) = basicSnapLevels
+        in  (fold [tagA, tagB, tagC, tagD ], SnapshotMetaData valA valB valC valD)
+  in  testCodecBuilder "SnapshotLabels" $ assembler <$> enumerateSnapshotLabel
+
+testCodecSnapshotTableType :: TestTree
+testCodecSnapshotTableType =
+  let assembler (tagB, valB) =
+        let (tagA, valA) = basicSnapshotLabel
+            (tagC, valC) = basicTableConfig
+            (tagD, valD) = basicSnapLevels
+        in  (fold [tagA, tagB, tagC, tagD ], SnapshotMetaData valA valB valC valD)
+  in  testCodecBuilder "SnapshotTables" $ assembler <$> enumerateSnapshotTableType
+
+testCodecTableConfig :: TestTree
+testCodecTableConfig =
+  let assembler (tagC, valC) =
+        let (tagA, valA) = basicSnapshotLabel
+            (tagB, valB) = basicSnapshotTableType
+            (tagD, valD) = basicSnapLevels
+        in  (fold [tagA, tagB, tagC, tagD ], SnapshotMetaData valA valB valC valD)
+  in  testCodecBuilder "SnapshotConfig" $ assembler <$> enumerateTableConfig
+
+testCodecSnapLevels :: TestTree
+testCodecSnapLevels =
+  let assembler (tagD, valD) =
+        let (tagA, valA) = basicSnapshotLabel
+            (tagB, valB) = basicSnapshotTableType
+            (tagC, valC) = basicTableConfig
+        in  (fold [tagA, tagB, tagC, tagD ], SnapshotMetaData valA valB valC valD)
+  in  testCodecBuilder "SnapshotLevels" $ assembler <$> enumerateSnapLevels
+
+testCodecBuilder :: TestName -> [(ComponentAnnotation, SnapshotMetaData)] -> TestTree
+testCodecBuilder tName metadata =
+  testGroup tName $ uncurry snapshotCodecTest <$> metadata
+
+type ComponentAnnotation = String
+
+{----------------
+Defaults used when the SnapshotMetaData sub-component is not under test
+----------------}
+
+basicSnapshotLabel :: (ComponentAnnotation, SnapshotLabel)
+basicSnapshotLabel = head enumerateSnapshotLabel
+
+basicSnapshotTableType :: (ComponentAnnotation, SnapshotTableType)
+basicSnapshotTableType = head enumerateSnapshotTableType
+
+basicTableConfig :: (ComponentAnnotation, TableConfig)
+basicTableConfig = ("T", defaultTableConfig)
+
+basicSnapLevels :: (ComponentAnnotation, SnapLevels RunNumber)
+basicSnapLevels = head enumerateSnapLevels
+
+{----------------
+Enumeration of SnapshotMetaData sub-components
+----------------}
+
+enumerateSnapshotLabel :: [(ComponentAnnotation, SnapshotLabel)]
+enumerateSnapshotLabel =
+  [ ("Bs", SnapshotLabel "UserProvidedLabel")
+  , ("Bn", SnapshotLabel "")
+  ]
```

Comment:
So I think I understand now that the `ComponentAnnotation` is used to name the specific enumerations, and this is reflected in the names of the golden files. Can you explain how you picked these labels? And the labels for the other enumerators below?